ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -359,8 +359,10 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
         -- Normally this will be the current block at the tip, but it may
         -- be the /previous/ block, if there were multiple slot leaders
         (prevPoint, prevNo) <- do
-          mPrev <- lift $ atomically $ prevPointAndBlockNo currentSlot <$>
-                     ChainDB.getCurrentChain chainDB
+          (cc, cBno) <- lift $ atomically $
+            (,) <$> ChainDB.getCurrentChain chainDB
+                <*> ChainDB.getTipBlockNo chainDB
+          let mPrev = prevPointAndBlockNo currentSlot cBno cc
```

Comment:
What about the following:
```haskell
          mPrev <- lift $ atomically $
            prevPointAndBlockNo currentSlot
                <$> ChainDB.getTipBlockNo chainDB
                <*> ChainDB.getCurrentChain chainDB
```
Because of laziness, we won't be doing the extra work in the transaction.

Reply:
Yes, let's do this.